framework/src/mediaplayer/mediaplayer.c

```diff
@@ -0,0 +1,581 @@
+	m->op = op;
+	m->state = MEDIA_STATE_CREATED;
+	m->type = type;
+	m->play_type = play_type;
+
+	if (type == MEDIA_TYPE_WAV || type == MEDIA_TYPE_PCM) {
+		if (op == MEDIA_OP_PLAYBACK) {
+			if (play_type == MEDIA_PLAY_TYPE_FILE) {
+				m->fd = open(path, O_RDONLY);
+				if (m->fd < 0) {
+					free(m);
+					return NULL;
+				}
+
+				if (m->type == MEDIA_TYPE_WAV) {
+					read_wav_header(m);
+				}
+				m->state = MEDIA_STATE_PLAYING;
+				return m;
+			} else if (play_type == MEDIA_PLAY_TYPE_NETWORK) {
+				// TODO: Network streaming support
+				free(m);
+				return NULL;
+			}
+			free(m);
+			return NULL;
+
+		} else if (op == MEDIA_OP_RECORD) {
+			if (play_type == MEDIA_PLAY_TYPE_FILE) {
+				m->fd = open(path, O_RDWR | O_CREAT | O_TRUNC);
+				if (m->fd < 0) {
+					free(m);
+					return NULL;
+				}
+				m->state = MEDIA_STATE_RECORDING;
+				return m;
+			} else if (play_type == MEDIA_PLAY_TYPE_NETWORK) {
+				// TODO: Network streaming support
+				free(m);
+				return NULL;
+			}
+			free(m);
+			return NULL;
+		}
+	}
+	free(m);
+	return NULL;
+}
+
+void media_close(media_t *m)
+{
+	if (m == NULL) {
+		return;
+	}
+	if (m->fd >= 0) {
+		close(m->fd);
+	}
+	free(m);
+	return;
+}
+
+
+int on_media_state_playing(struct pcm *pcmout, media_t *m, char *buffer, unsigned int buffer_size)
+{
+	int readed;
+	int remain;
+	int ret;
+	readed = read(m->fd, buffer, buffer_size);
+	if (readed) {
+		remain = pcm_bytes_to_frames(pcmout, readed);
+		while (remain > 0) {
+			ret = pcm_writei(pcmout, buffer + readed - pcm_frames_to_bytes(pcmout, remain), remain);
+			if (ret <= 0) {
+				return MEDIA_ERROR_PCM_WRITE;
+			} else {
+				remain -= ret;
+			}
+		}
+	}
+
+	return readed;
+}
+
+int on_media_state_recording(struct pcm *pcmin, media_t *m, char *buffer, unsigned int buffer_size)
+{
+	int readed;
+	int remain;
+	int ret;
+
+	readed = pcm_readi(pcmin, buffer, pcm_bytes_to_frames(pcmin, buffer_size));
+	if (readed) {
+		remain = pcm_frames_to_bytes(pcmin, readed);
+		while (remain > 0) {
+			ret = write(m->fd, buffer + pcm_frames_to_bytes(pcmin, readed) - remain, remain);
+			if (ret < 0) {
+				return MEDIA_ERROR_WRITE_TO_FILE;
+			} else {
+				remain -= ret;
+			}
+		}
+	}
+	return readed;
+}
+
+/***
+ * Thread routine
+ */
+media_result_t audio_playing(void *args)
+{
+	char *buffer;
+	unsigned int buffer_size;
+	struct pcm *pcmout;
+	struct pcm_config config;
+	dq_entry_t *node;
+	int ret;
+#ifndef CONFIG_AUDIO_MULTI_CARD
+	struct pcm *pcmin;
+	bool playing = false;
+	bool recording = false;
+	buffer = NULL;
+	buffer_size = 0;
+	pcmout = NULL;
+	pcmin = NULL;
+#endif
+	// Configuration
+	config.channels = 2;
+	config.rate = 16000;
+	config.period_size = CONFIG_AUDIO_BUFFER_NUMBYTES;
+	config.period_count = CONFIG_AUDIO_NUM_BUFFERS;
+	config.format = PCM_FORMAT_S16_LE;
+	config.start_threshold = 0;
+	config.stop_threshold = 0;
+	config.silence_threshold = 0;
+
+#ifdef CONFIG_AUDIO_MULTI_CARD
+	pcmout = pcm_open(0, 0, PCM_OUT, &config);
```

Comment:
Below is more stable.
buffer = malloc();
if (buffer == NULL) {
    return ERROR;
}
pcm = pcm_open();
if (pcm == NULL) {
   return ERROR;
}